### PR TITLE
Tidy up computation of import paths

### PR DESCRIPTION
**Pull request by EdSchouten**
Right now you see that jsonnet_to_json() and jsonnet_to_json_test()    implicitly call jsonnet with "-J .". This means that if these rules are    used in combination with libraries that are declared in the root module,    everything works as expected. But as soon as libraries are used from    other modules, importing them becomes more tedious.
    
The goal of this change is to ensure that if a Jsonnet project can be    built within the root module, that module can also safely be used as a    child module. We solve this by automatically adding the workspace root    to the set of import paths for which one or more source files exist.    Furthermore, by considering the root of every source file, we no longer    need to use bin_dir and genfiles_dir.
    
Fixes: #44
Fixes: #86
Fixed: #139
Fixes: #154
Fixes: #178

### CODE REVIEW

File: jsonnet/jsonnet.bzl [hunk 1]
```diff
@@ -210,12 +219,10 @@ def _jsonnet_to_json_impl(ctx):
         [
             "set -e;",
             toolchain.jsonnet_path,
-        ] + ["-J %s" % im for im in _add_prefix_to_imports(ctx.label, ctx.attr.imports)] +
-        ["-J %s" % im for im in depinfo.imports.to_list()] + [
-            "-J .",
-            "-J %s" % ctx.genfiles_dir.path,
-            "-J %s" % ctx.bin_dir.path,
```
Comment (agotsis) [inline jsonnet/jsonnet.bzl:217]: @EdSchouten - the removal of the bin_dir is a divergence in behavior from the past. Do you feel that this is more correct?

(I'm encountering an issue related to this in a large repo - but wanting to understand if the change is intentional)

Comment (EdSchouten) [reply jsonnet/jsonnet.bzl:217]: The bindir should still be added for generated source files, through file.root.path.

Comment (EdSchouten) [reply jsonnet/jsonnet.bzl:217]: Could you make a small reproduction case?

**Comment by frimik** [comment]
Something here seems to have broken imports for all our rules relying on `imports` being populated from a couple of core `deps`. Now despite "top-level" deps having for example; `imports = ["lib", "vendor"]` the only thing showing up in the jsonnet command are the two duplicated `.` arguments, like: `-J . -J .`.


`lib/kube-libsonnet/BUILD`:
```
jsonnet_library(
    name = "kube",
    srcs = ["kube.libsonnet"],
    visibility = ["//visibility:public"],
    deps = ["//:imports"], # this doesn't work anymore ...
    imports = ["../"] # this had to be added now
)
```

With the explicit `imports = ["../"] in this library, it works.

The trickery I did in previous versions was a top-level `//:imports` rule which was a dep of the other libs, see example:

```python
jsonnet_library(
    name = "imports",
    imports = [
        "./lib",
        "./vendor",
    ],
)

jsonnet_library(
    name = "k",
    srcs = glob([
        "vendor/github.com/jsonnet-libs/k8s-libsonnet/1.25/**/*.libsonnet",
        "lib/k/**/*.libsonnet",
    ]) + ["lib/k.libsonnet"],
    visibility = ["//visibility:public"],
    deps = [
        ":imports",
        ":ksonnet-util",
    ],
)
```

Why? This was to make other jsonnet tools and bazel able to co-exist. Many of those tools, and editor configurations expect the import paths to be `-J vendor -J lib` to have good mechanisms for both imports and overrides available.

**Comment by EdSchouten** [comment]
The reason this construct no longer works is becasue `jsonnet_library(name = "imports")` does not have any sources. Without sources, the rule isn't able to determine whether it should add include paths for the source directory or the output directory.

Instead of using that dummy `jsonnet_library(name = "imports")`, why can't you add `imports = []` to `jsonnet_library(name = "k")` directly? If you want to reuse these import paths between various targets in that same package, keep in mind that you can always put those in a constant `MY_IMPORTS = ["lib", "vendor"]`.

**Pull request merged**